src/backend/app.py

```diff
@@ -11,25 +16,123 @@
-    if not result.status:
-        return result.message, 400
-    return result.message, 200
+    json_form = request.get_json(force=True)
+
+    if isinstance(json_form, dict):
+        username = json_form.get("username", "")
+        email = json_form.get("email", "")
+        password = json_form.get("password", "")
+        phone = json_form.get("phone", "")
+        result = user_login.register(username, email, password, phone)
+        if not result.status:
+            return result.message, 400
+        return result.message, 201
+    return "", 400
+
+
+@app.route("/", methods=["POST", "GET"])
+@app.route("/main", methods=["POST", "GET"])
+def mainpage():
+    """Handle mainpage requests"""
+    mainpage_obj = MainPage()
+    if request.method == "POST":
+        return mainpage_post(mainpage_obj)
+
+    args = request.args
+    return mainpage_get(mainpage_obj, args)
+
+
+def mainpage_get(mainpage_obj: MainPage, args: MultiDict):
+    """Helper for mainpage get requests"""
+    action_type = namedtuple(
+        "action_type", ["is_search", "is_populate", "is_review", "is_pictures"]
+    )
+    action = action_type(
+        args.get("search", default=False, type=bool),
+        args.get("populate", default=False, type=bool),
+        args.get("review", default=False, type=bool),
+        args.get("pictures", default=False, type=bool),
+    )
+
+    params = namedtuple(
+        "params", ["num_apts", "apt_id", "search_query", "price_sort", "rating_sort"]
+    )
+    param = params(
+        args.get("numApts", type=int),
+        args.get("aptId", type=int),
+        args.get("searchQuery", type=str),
+        args.get("priceSort", type=int),
+        args.get("ratingSort", type=int),
+    )
+
+    query_result = ""
+    if action.is_search is True and param.search_query is not None:
+        apts = mainpage_obj.search_apartments(param.search_query)
+        apts_dict = [dataclasses.asdict(apt) for apt in apts]
+        query_result = json.dumps(apts_dict)
+
+    elif action.is_populate is True and param.num_apts is not None:
+        apts = []
+        if param.price_sort is not None and param.rating_sort is not None:
+            apts = mainpage_obj.apartments_sorted(
+                param.num_apts, param.price_sort, param.rating_sort
+            )
+        else:
+            apts = mainpage_obj.apartments_default(param.num_apts)
+        apts_dict = [dataclasses.asdict(apt) for apt in apts]
+        query_result = json.dumps(apts_dict)
+
+    elif action.is_review is True and param.apt_id is not None:
+        reviews = mainpage_obj.get_apartments_reviews(param.apt_id)
+        reviews_dict = [dataclasses.asdict(review) for review in reviews]
+        query_result = json.dumps(reviews_dict)
+
+    elif action.is_pictures is True and param.apt_id is not None:
+        query_result = json.dumps(mainpage_obj.get_apartments_pictures(param.apt_id))
+
+    if len(query_result) != 0:
```

Comment:
We still want to sort if a button was selected but the query is empty, right?

Reply:
Wdym a button was selected but the query is empty?

Reply:
You just searched for something and now want to sort it

Reply:
> You just searched for something and now want to sort it

Those are separate functionalities, with separate gets requests.

Reply:
Currently the search bar just displays some apartment names, it does nothing else.